cartographer/mapping/internal/constraints/constraint_builder_2d.cc

```diff
@@ -258,50 +266,41 @@ void ConstraintBuilder2D::ComputeConstraint(
   }
 }
 
-void ConstraintBuilder2D::FinishComputation(const int computation_index) {
+void ConstraintBuilder2D::RunCallback() {
   Result result;
   std::unique_ptr<std::function<void(const Result&)>> callback;
   {
     common::MutexLocker locker(&mutex_);
-    if (--pending_computations_[computation_index] == 0) {
-      pending_computations_.erase(computation_index);
-    }
-    if (pending_computations_.empty()) {
-      CHECK_EQ(submap_queued_work_items_.size(), 0);
-      if (when_done_ != nullptr) {
-        for (const std::unique_ptr<Constraint>& constraint : constraints_) {
-          if (constraint != nullptr) {
-            result.push_back(*constraint);
-          }
-        }
-        if (options_.log_matches()) {
-          LOG(INFO) << constraints_.size() << " computations resulted in "
-                    << result.size() << " additional constraints.";
-          LOG(INFO) << "Score histogram:\n" << score_histogram_.ToString(10);
-        }
-        constraints_.clear();
-        callback = std::move(when_done_);
-        when_done_.reset();
+    CHECK(when_done_ != nullptr);
+    for (const std::unique_ptr<Constraint>& constraint : constraints_) {
+      if (constraint != nullptr) {
+        result.push_back(*constraint);
       }
     }
+    if (options_.log_matches()) {
+      LOG(INFO) << constraints_.size() << " computations resulted in "
+                << result.size() << " additional constraints.";
+      LOG(INFO) << "Score histogram:\n" << score_histogram_.ToString(10);
+    }
+    constraints_.clear();
+    callback = std::move(when_done_);
+    when_done_.reset();
     kQueueLengthMetric->Set(constraints_.size());
   }
-  if (callback != nullptr) {
-    (*callback)(result);
-  }
+  (*callback)(result);
 }
 
 int ConstraintBuilder2D::GetNumFinishedNodes() {
   common::MutexLocker locker(&mutex_);
-  if (pending_computations_.empty()) {
-    return current_computation_;
-  }
-  return pending_computations_.begin()->first;
+  return num_finished_nodes_;
```

Comment:
It's a shame to take a lock for something so trivial like this counter. Let's turn this guy into a std::atomic

Reply:
Added a todo.